### PR TITLE
uORB::Publication: automatically obtain ORB_QUEUE_LENGTH from the structure

### PR DESCRIPTION
**Pull request by ShawnFeng0**
Use the C++ template technique to get the `ORB_QUEUE_LENGTH` in the structure, and return the default value if it does not contain `ORB_QUEUE_LENGTH`.

In this way, there is no need to remember whether a topic has a queue, `uORB::Publication` will automatically configure the queue size according to the topic in the *.msg file.

### CODE REVIEW

**Comment by dagar** [comment]
There's a bit of a hack on destruction to be aware of.
https://github.com/PX4/Firmware/blob/b14d0e43265b25a49905c9d543756ed442cbbb2c/src/modules/uORB/Publication.hpp#L67-L71

This is because of standalone uses publishing a vehicle_command by stack allocating uORB::PublicationQueued and we need the vehicle_command to still exist. Maybe we need a slightly different approach for this use case?

**Comment by ShawnFeng0** [comment]
It may be feasible to put this destructor in Publication and PublicationMulti processing, how about it?
PublicationBase does not perform any functions, but only provides public interfaces and data.

But I think this use case affects the consistency of architecture performance, maybe we should delete this use case later.

I think no matter whether the broadcaster exists or not, the subscribers should be able to continue to get the unobtained data and make the system behave consistently.

**Comment by dagar** [comment]
> I think no matter whether the broadcaster exists or not, the subscribers should be able to continue to get the unobtained data and make the system behave consistently.

I think what could work is leaving it be (skipping unadvertise) if any data was actually published. If the uORB::DeviceNode never published anything and doesn't have any subscribers then it should be deleted on unadvertise.

Later the entire uORB shutdown/cleanup story needs to be reviewed as it's largely been ignored due to most people turning off their drones by pulling power abruptly.

**Comment by ShawnFeng0** [comment]
Of course, the data should be kept in DeviceNode.

My thoughts are a little different:

`Unadvertise` only reduces the number of publishers in DeviceNode. Without affecting data acquisition. `Unadvertise` has no practical meaning.

We don't care whether there is a queue, we store the number of publishers/subscribers and queue data, and implement management in DeviceNode based on these data.

For example, this situation:
After the publisher publishes the data, it is deconstructed. We should reduce the number of publishers by 1, but if the number of subscribers is greater than 0, we will not process DeviceNode. Subscribers can continue to get data.

We only delete DeviceNode when the number of publishers and subscribers are both 0.

**Comment by ShawnFeng0** [comment]
By the way, in the current design, if a DeviceNode is published by multiple publishers, as long as one of the publishers is destroyed, the DeviceNode will be marked as unadvertised. (Need to record the number of publishers to solve.)

**Comment by ShawnFeng0** [comment]
Can this modification be merged? The unadvertise related issues mentioned above will be a new pull request.

File: src/modules/uORB/Publication.hpp [hunk 1]
```diff
@@ -48,6 +48,24 @@
...
 {
 
+template <typename U> class DefaultQueueSize
+{
+private:
+	template <typename T>
+	static constexpr uint8_t get_queue_size(decltype(T::ORB_QUEUE_LENGTH) *)
+	{
+		return T::ORB_QUEUE_LENGTH;
+	};
```
Comment (bkueng) [inline src/modules/uORB/Publication.hpp:58]: Extra semicolon

File: src/modules/uORB/Publication.hpp [hunk 2]
```diff
@@ -48,6 +48,24 @@
...
+
+	template <typename T> static constexpr uint8_t get_queue_size(...)
+	{
+		return 1;
+	};
```
Comment (bkueng) [inline src/modules/uORB/Publication.hpp:63]: Same

File: src/modules/uORB/Publication.hpp (hunk 1 above)
Comment (ShawnFeng0) [reply src/modules/uORB/Publication.hpp]: Modified.

File: src/modules/uORB/Publication.hpp (hunk 2 above)
Comment (ShawnFeng0) [reply src/modules/uORB/Publication.hpp]: Modified.

**Review by bkueng — APPROVED** [review]
Thanks

**Comment by lukegluke** [comment]
This commits leads to specific factor.

[I'm working with custom board and project based on PX4]
I had LedsSimple driver that works in the same way as rgbled and calls `_led_controller.update` in Run().

in rcS right after this drivers starts I call `led_control` cmd three times in a row:
```
led_control off -p 0                 #turn off all leds
led_control on -l 0 -p 0             #turn on one led with 0 priority
led_control blink -n 3 -s fast -p 2  #blink all leds three times
```
before (when PublicationQueued was used) after blinking there is on led stay lighting as expected, now it doesn't.
Debugging showed that `led_control` cmd has time to publish all three orb messages in queue, before driver reaches `_led_controller.update` call and led controller subscribes to topic - so generation is already 3 for this moment and it reads only last message - blinking.
So in my case adding immediate `_led_control_sub.subscribe()` in `LedController` constructor fixes the problem.

But as I see the PX4 boards probably don't have such code that would call led_control/tune_control or other related commands in row on boot, right? Anyway just a warning.

**Comment by ShawnFeng0** [comment]
That is due to this https://github.com/PX4/PX4-Autopilot/pull/15709
After subscribing, current subscribers can only get one old data of before subscription. https://github.com/PX4/PX4-Autopilot/blob/66edc822febf6f73b096111c8acfab9ef9a742e5/src/modules/uORB/uORBDeviceNode.cpp#L547-L557

**Comment by lukegluke** [comment]
Yes, I saw this. And it is also due to Subscription will not subscribe until topic will be advertise:
https://github.com/PX4/PX4-Autopilot/blob/66edc822febf6f73b096111c8acfab9ef9a742e5/src/modules/uORB/Subscription.cpp#L58

**Pull request merged**